halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/config/v1/HalconfigParser.java

```diff
@@ -68,6 +70,9 @@
 
   private boolean useBackup = false;
 
+  // @LocalFile fields whose file path is relative
+  private Map<Node, Map<Field, String>> originalRelativePaths = new ConcurrentHashMap<>();
```

Comment:
Two broken things here:
1) Adding state here is weird because a `@Component` is a singleton, so this is now essentially a global variable... everyone who injects a `HalconfigParser` now shares this map.
2) this data is specific to a given `Halconfig`, but you don't tie it to any specific `Halconfig`, you just apply it to whatever object is handed to you

Reply:
Would it be better to move this back to `Node`? 
Essentially, paths are tied to specific `Node` instances (for example, a `Halconfig` may have multiple Kubernetes accounts, each one with its own individual `@LocalFile` for the kubeconfig), that's why I'm keying by `Node` instead of `Halconfig`.

Reply:
Sorry for the essay and also I 100% realize most of these problems existed before you so please don't think I'm picking on you.

But anyway: where should this code and this temporary data live?

Not in `Node` or `Halconfig`. These are data objects and by any modern coding standard, they should be immutable. They certainly shouldn't have transient data or methods that mutate them in seemingly arbitrary ways. And anyway, 99% of the users of `Halconfig` don't care one bit about transforming absolute paths, so it probably shouldn't live there.

Not in `HalconfigParser`. It's a singleton, so shouldn't have state, as mentioned. But also, it's right there in the name: it's a parser, there's no reason it should also have code to modify data.

Maybe in `BackupService`. This code and data exists _solely_ for use by backups. So this is the first logical place I see where it might live. Putting the code here definitely would make sense. But it's also a singleton, so shouldn't have state.

The state should ideally live in a `BackupScope` object or something. That's how we typically model something like this. Then you pass the scope into the `BackupService`. But if you do that, you'll find you have nowhere to store the scope between calls to `BackupService#create` and `BackupService#restore`. And that's your first clue why this implementation won't work: you have no guaranteee that someone didn't restart halyard between those two cals and lose your state! That's why `{%halconfig-dir%}` was there.

But I think you just don't need any of this anyway. What this is doing is making it so when you back up `/home/german/.hal/mykey.json`, it gets restored on my box as `/home/plumpy/.hal/mykey.json`. But if you're going to make relative paths work anyway, who cares if we just restore it instead to `mykey.json`. Why bother restoring the absolute path at all?

Reply:
These are valid points. As for what's the purpose of this in-memory state, when we can just go away using relative paths in hal `config` file: there is some code, specially validators, that during the validation process try to load these files from file system with regular `IOUtils.toByteArray(new FileInputStream(path))`. This code doesn't work with relative file paths because it's not resolving relative to hal config home, so here I'm converting all paths to absolute when the file is loaded from disk. After changes are done to `Halconfig` structure and it's written back to disk, I need to restore the original relative paths.

So this flow of converting paths to absolute when `Halconfig` is loaded, and then write them back to disk as relative, is not just for doing `hal config backup/restore`, is for any code that wants to load a file to continue working, for example, any `hal config provider kubernetes` command that wants to load kubeconfig files.

I guess a better solution would be to address the root cause, and when any code tries to read a local file, append hal config home if the path is relative just in that moment. In this way there's no need for in-memory state or for path manipulation when loading or saving `Halconfig`. The downside I see is that this solution may be a bit fragile on maintenance, as anyone that is not aware that there's a helper function for loading files, could easily bypass it by doing plain `new File()` or something else. Anyway, it would be worth implementing.